mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala

```diff
@@ -0,0 +1,472 @@
+      (instance: Instance, model: WeightedLeastSquaresModel) => {
+        val eta = model.predict(instance.features)
+        val mu = familyLink.fitted(eta)
+        val z = familyLink.adjusted(instance.label, mu, eta)
+        val w = familyLink.weights(mu) * instance.weight
+        (z, w)
+      }
+    }
+
+    val optimizer = new IterativelyReweightedLeastSquares(initialModel, reweightFunc,
+      $(fitIntercept), $(regParam), $(maxIter), $(tol))
+
+    val irlsModel = optimizer.fit(instances)
+
+    val model = copyValues(new GeneralizedLinearRegressionModel(uid,
+      irlsModel.coefficients, irlsModel.intercept).setParent(this))
+    model
+  }
+
+  @Since("2.0.0")
+  override def copy(extra: ParamMap): GeneralizedLinearRegression = defaultCopy(extra)
+}
+
+@Since("2.0.0")
+object GeneralizedLinearRegression {
+
+  /** Set of families that GeneralizedLinearRegression supports */
+  private[ml] val supportedFamilies = Set("gaussian", "binomial", "poisson", "gamma")
+
+  /** Set of links that GeneralizedLinearRegression supports */
+  private[ml] val supportedLinks = Set("identity", "log", "inverse", "logit", "probit",
+    "cloglog", "sqrt")
+
+  /** Set of family and link pairs that GeneralizedLinearRegression supports */
+  private[ml] val supportedFamilyLinkPairs = Set(
+    "gaussian" -> "identity", "gaussian" -> "log", "gaussian" -> "inverse",
+    "binomial" -> "logit", "binomial" -> "probit", "binomial" -> "cloglog",
+    "poisson" -> "log", "poisson" -> "identity", "poisson" -> "sqrt",
+    "gamma" -> "inverse", "gamma" -> "identity", "gamma" -> "log"
+  )
+}
+
+/**
+ * :: Experimental ::
+ * Model produced by [[GeneralizedLinearRegression]].
+ */
+@Experimental
+@Since("2.0.0")
+class GeneralizedLinearRegressionModel private[ml] (
+    @Since("2.0.0") override val uid: String,
+    @Since("2.0.0") val coefficients: Vector,
+    @Since("2.0.0") val intercept: Double)
+  extends RegressionModel[Vector, GeneralizedLinearRegressionModel]
+  with GeneralizedLinearRegressionParams {
+
+  private lazy val familyLink = $(family) match {
+    case "gaussian" => Gaussian($(link))
+    case "binomial" => Binomial($(link))
+    case "poisson" => Poisson($(link))
+    case "gamma" => Gamma($(link))
+  }
+
+  override protected def predict(features: Vector): Double = {
+    val eta = BLAS.dot(features, coefficients) + intercept
+    familyLink.fitted(eta)
+  }
+
+  @Since("2.0.0")
+  override def copy(extra: ParamMap): GeneralizedLinearRegressionModel = {
+    copyValues(new GeneralizedLinearRegressionModel(uid, coefficients, intercept), extra)
+      .setParent(parent)
+  }
+}
+
+/**
+ * A description of the error distribution and link function to be used in the model.
+ * @param link a link function instance
+ */
+private[ml] abstract class Family(val link: Link) extends Serializable {
+
+  /** Initialize the starting value for mu. */
+  def initialize(y: Double, weight: Double): Double
+
+  /** The variance of mu to its mean. */
```

Comment:
I think this comment is a bit confusing. This is really the variance of the endogenous variable, given the expected value `mu`.